src/main.py

```diff
@@ -1,79 +1,53 @@
-from flask import Flask, render_template, abort
+from flask import Flask, render_template
 from flask_talisman import Talisman
 from csp import csp
+from validate_i18n import validate_i18n
 import logging
 
 app = Flask(__name__)
 Talisman(app,
          content_security_policy=csp,
          content_security_policy_nonce_in=['script-src'])
-
-DEFAULT_YEAR = '2019'
-DEFAULT_LANG = 'en'
-SUPPORTED_YEARS = {
-    '2019': ('en', 'ja')
-}
+logging.basicConfig(level=logging.DEBUG)
 
 
 @app.route('/')
 @app.route('/<lang>/')
-def index(lang=DEFAULT_LANG):
-    validate(lang)
-
+@validate_i18n
+def index(year, lang):
```

Comment:
What happens if we remove the year from the parameters?

Reply:
This was the only compromise for the solution. If the year is not in the parameters, it will throw some kind of key exception. While this was a simplifying assumption (I thought all pages made sense to be versioned by year), if it doesn't hold then we can partially update `**kwargs` only with fields that were originally present.

Reply:
The home page is year-agnostic since it is the landing page for the current and all earlier editions, so it'd be nice if we can remove the param. I'm surprised that it's required because `args` should handle any input.

Reply:
This is my only lingering concern, otherwise this PR LGTM.

This PR is a big improvement to both code quality and functionality. Thanks so much for working on it!

Reply:
I used the `inspect` module to get the args that the route will accept, which will allow us to also optionally update chapter going forward.

```py
         accepted_args = inspect.getargspec(func).args

         lang, year = validate_lang_and_year(lang, year)

         if 'lang' in accepted_args:
            kwargs.update({'lang': lang})

         if 'year' in accepted_args:
            kwargs.update({'year': year})
```

It was the right change to make, thanks for insisting on it!

(Also: it's a pleasure. :) )